### PR TITLE
Minor: improve `join` / `join_on` docs

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/arrow-datafusion/pull/7805~

## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/pull/7805 and https://github.com/apache/arrow-datafusion/issues/7766

## Rationale for this change

The capabilities of datafusion with `joins` could be better documented, so let's fix that. 

## What changes are included in this PR?

Update documentation comments

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -650,8 +659,15 @@ impl LogicalPlanBuilder {
...
-    /// let exprs = vec![col("left.a").eq(col("right.a")), col("left.b").not_eq(col("right.b"))];
+    /// // Form the expression `(left.a != right.a)` AND `(left.b != right.b)`
+    /// let exprs = vec![
+    ///     col("left.a").eq(col("right.a")),
+    ///     col("left.b").not_eq(col("right.b"))
+    ///  ];
     ///
+    /// // Perform the equivalent of `left INNER JOIN right ON (a != a2 AND b != b2)`
+    /// // finding all pairs of rows from `left` and `right` where
+    /// // where `a != a2` and `b != b2`.
```
Comment (jackwener) [inline datafusion/expr/src/logical_plan/builder.rs:670]:
```suggestion
    /// // where `a = a2` and `b != b2`.
```

**Review by jackwener — APPROVED** [review]
Thanks @alamb

File: datafusion/expr/src/logical_plan/builder.rs (hunk 1 above)
Comment (alamb) [reply datafusion/expr/src/logical_plan/builder.rs:670]: Good catch -- thanks @jackwener

**Comment by alamb** [comment]
Thanks for the review @jackwener

**Pull request merged**